nodejs_container/index.js

```diff
@@ -23,26 +69,97 @@ Container.prototype.call = function (id, zome, trait, fn, params) {
     return result;
 }
 
+Container.prototype.callWithPromise = function (...args) {
+    const promise = new Promise((fulfill, reject) => {
+        this.register_callback(() => fulfill(result))
+    })
+    const result = this.call(...args)
+    return [result, promise]
+}
+
+Container.prototype.callSync = function (...args) {
+    const [result, promise] = this.callWithPromise(...args)
+    return promise
+        .catch(err => console.error("Error with scenario test system: ", err))
+        .then(() => { return result })
+}
+
 // Convenience function for making an object that can call into the container
 // in the context of a particular instance. This may be temporary.
 Container.prototype.makeCaller = function (agentId, dnaPath) {
-  const instanceId = agentId + '::' + dnaPath
+  const instanceId = makeInstanceId(agentId, dnaPath)
   return {
     call: (zome, cap, fn, params) => this.call(instanceId, zome, cap, fn, params),
     agentId: this.agent_id(instanceId)
   }
 }
 
-const Config = {
-    agent: name => ({ name }),
-    dna: (path) => ({ path }),
-    instance: (agent, dna, name) => {
-        if (!name) {
-            name = agent.name
+Container.withInstances = function (instances, opts=defaultOpts) {
+    const config = Config.container(instances, opts)
+    return new Container(config)
+}
+
+/////////////////////////////////////////////////////////////
+
+class Scenario {
+    constructor(instances, opts=defaultOpts) {
+        this.instances = instances
+        this.opts = opts
+    }
+
+    static setTape(tape) {
+        Scenario._tape = tape
+    }
+
+    /**
+     * Run a test case, specified by a closure:
+     * (stop, {instances}) => { test body }
+     * where `stop` is a function that ends the test and shuts down the running Container
+     * and the `instances` is an Object of instances specified in the config, keyed by "name"
+     * (name is the optional third parameter of `Config.instance`)
+     *
+     * e.g.:
+     *      scenario.run(async (stop, {alice, bob, carol}) => {
+     *          const resultAlice = await alice.callSync(...)
+     *          const resultBob = await bob.callSync(...)
+     *          assert(resultAlice === resultBob)
+     *          stop()
+     *      })
+     */
+    run(fn) {
+        const container = Container.withInstances(this.instances, this.opts)
+        container.start()
+        const callers = {}
+        this.instances.forEach(instance => {
+            const id = makeInstanceId(instance.agent.name, instance.dna.name)
```

Comment:
I‘m not fully on board with hard-coding the instance ID from agent and DNA. I had this case recently where I wanted to have multiple (separate) instances in separate tests (instances that wouldn’t even communicate). I had to add another agent because if this.

I guess you would say with these changes I could/should create separate scenarios for such a case?

Reply:
... and I guess you‘d be right. It’s just that I think it might be worthwhile to also expose the ability to explicitly set the instance id.. if you know what you‘re doing..

Reply:
@lucksus I agree, actually I'm not sure if the "name" part of the Config.instance is or isn't used as the instance ID, but it probably should be. Happy to make that change in a future revision.

Also the thing about scenario tests now is that *every* test is separate from each other test, no instances communicate across tests. So would you still need to create multiple instances as you describe in that case?